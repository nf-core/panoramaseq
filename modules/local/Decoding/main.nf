//decode Using 2 gpus and high resolution array
process Decode_batch {
    tag "$meta.id" 
    label "gpu_process"
    conda "${moduleDir}/environment.yml"
    container 'docker://pytorch/pytorch:2.7.1-cuda11.8-cudnn9-runtime'
    // publishDir "${params.outdir}/${index_step}_demux", mode: 'copy', overwrite: true 
    
    input:
    tuple val(meta), path(reads)
    
    output:
    tuple val(meta), path("*.fastq.gz"), emit: reads
    
    when:
    task.ext.when == null || task.ext.when

    script:
    // Defensive debug print and assertion
    println "[Decode_batch] meta.id: ${meta.id}, N_barcodes: ${meta.N_barcodes}, sample_size: ${meta.sample_size}"
    assert meta.N_barcodes != null : "N_barcodes is null for sample ${meta.id}"
    assert meta.N_barcodes > 0 : "N_barcodes is not positive for sample ${meta.id}"

    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def gpus = params.gpus ?: 2
    def cmds = []
    for (int i = 1; i <= gpus; i++) {
        int cuda = (i-1) % gpus
        cmds << "Demux2_batch.py -R1 ${reads[0]} -R2 ${reads[1]} -N ${meta.N_barcodes} -M ${meta.len_barcode} -B ${meta.barcode_file} --Ntriage ${meta.Ntriage} --Nthresh ${meta.Nthresh} --out-prefix ${meta.id} -C ${cuda} --Seg_id ${i}/${gpus} &"
        cmds << "pid${i-1}=\$!"
    }
    def waits = (0..<gpus).collect { "wait \$pid${it}" }.join("\n")
    def catR1 = "cat *${gpus}_R1.fastq.gz > ${prefix}.R1.fastq.gz"
    def catR2 = "cat *${gpus}_R2.fastq.gz > ${prefix}.R2.fastq.gz"
    def rm = "rm -f *${gpus}_R1.fastq.gz *${gpus}_R2.fastq.gz"
    """
    echo CUDA_VISIBLE_DEVICES=\${CUDA_VISIBLE_DEVICES:-}
    python -c "import torch; print('torch.cuda.is_available():', torch.cuda.is_available())"
    ${cmds.join("\n")}
    ${waits}
    ${catR1}
    ${catR2}
    ${rm}
    """
}