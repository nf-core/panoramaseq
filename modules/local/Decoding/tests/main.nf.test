nextflow_process {

    name "Test Process Decode_batch"
    script "../main.nf"
    process "Decode_batch"
    tag "modules"
    tag "modules_local"
    tag "decode_batch"

    test("decode batch stub test") {

        options "-stub"

        when {
            params {
                outdir = "$outputDir"
                panoramaseq_testdata_base_path = "https://raw.githubusercontent.com/francops1722/test-datasets/panoramaseq/testdata/"
            }
            process {
                """
                input[0] = [
                    [ 
                        id: 'test_decode', 
                        single_end: false,
                        N_barcodes: 5,
                        len_barcode: 36,
                        Ntriage: 3,
                        Nthresh: 2
                    ],
                    [
                        file(params.panoramaseq_testdata_base_path + 'decode/S09.umi_extract_1.fastq.gz', checkIfExists: true),
                        file(params.panoramaseq_testdata_base_path + 'decode/S09.umi_extract_2.fastq.gz', checkIfExists: true)
                    ]
                ]
                input[1] = file(params.panoramaseq_testdata_base_path + 'barcodes_coords.csv', checkIfExists: true)
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert process.out.reads != null },
                { assert process.out.reads.get(0).get(1).size() == 2 },
                { assert snapshot(process.out.versions).match() }
            )
        }
    }

    test("decode batch real processing") {

        when {
            params {
                outdir = "$outputDir"
                panoramaseq_testdata_base_path = "https://raw.githubusercontent.com/francops1722/test-datasets/panoramaseq/testdata/"
                gpus = 1  // Use only 1 GPU for testing
            }
            process {
                """
                input[0] = [
                    [ 
                        id: 'S09_real', 
                        single_end: false,
                        N_barcodes: 34500,  // Reduced for faster testing
                        len_barcode: 36,
                        Ntriage: 100,     // Reduced for faster testing
                        Nthresh: 9
                    ],
                    [
                        file(params.panoramaseq_testdata_base_path + 'decode/S09.umi_extract_1.fastq.gz', checkIfExists: true),
                        file(params.panoramaseq_testdata_base_path + 'decode/S09.umi_extract_2.fastq.gz', checkIfExists: true)
                    ]
                ]
                input[1] = file(params.panoramaseq_testdata_base_path + 'barcodes_coords.csv', checkIfExists: true)
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert process.out.reads != null },
                { assert process.out.reads.get(0).get(1).size() == 2 },
                { assert process.out.reads.get(0).get(1).get(0) ==~ ".*\\.R1\\.fastq\\.gz" },
                { assert process.out.reads.get(0).get(1).get(1) ==~ ".*\\.R2\\.fastq\\.gz" },
                { assert process.out.versions != null },
                { assert snapshot(process.out.versions).match("versions_real") }
            )
        }
    }
}
