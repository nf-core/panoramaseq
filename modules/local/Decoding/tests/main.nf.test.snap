{
    "decode batch stub test": {
        "content": [
            [
                "versions.yml:md5,449ef8004e3b7c713d514a3610a680ad"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-06-24T10:26:19.771131009"
    },
    "versions_real": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-06-24T12:57:59.848474355"
    }
}